File: Excel/400-499/490/490 Challenge.R
library(tidyverse)
library(readxl)

path = "Excel/490 - Fill Down.xlsx"

input = read_xlsx(path, range = "A1:B18")
test  = read_xlsx(path, range = "C1:C18")

result = input %>%
  fill(`Level 1`, .direction = "down") %>%
  group_by(group = cumsum(`Level 1` != lag(`Level 1`, default = first(`Level 1`))) + 1) %>%
  mutate(
    nr1 = row_number(),
    L2 = !is.na(`Level 2`) & nr1 != 1,
    L2_n2 = ifelse(L2, cumsum(L2), 0),
    `Answer Expected` = as.numeric(paste0(group, ".", L2_n2))
  ) %>%
  ungroup() %>%
  select(`Answer Expected`)

identical(result, test) 
#> [1] TRUE

File: Excel/400-499/490/490 Challenge.py
import pandas as pd

path = "490 - Fill Down.xlsx"
input = pd.read_excel(path, usecols="A:B")
test = pd.read_excel(path, usecols="C:C")

result = input.copy()
result["Level 1"].fillna(method="ffill", inplace=True)
result["group"] = (result["Level 1"] != result["Level 1"].shift()).cumsum()
result["nr1"] = result.groupby("group").cumcount() + 1
result["L2"] = (~result["Level 2"].isna()) & (result["nr1"] != 1)
result["L2_n2"] = result.groupby("group")["L2"].cumsum()
result.loc[~result["L2"], "L2_n2"] = 0
result["Answer Expected"] = result["group"].astype(str) + "." + result["L2_n2"].astype(str)
result = result[["Answer Expected"]].astype("float64").reset_index(drop=True)

print(result.equals(test)) # True